components/nimbus/src/enrollment.rs

```diff
@@ -696,6 +802,66 @@ fn map_enrollments(enrollments: &[ExperimentEnrollment]) -> HashMap<String, &Exp
     map_enrollments
 }
 
+/// Take a list of enrollments and a map of experiments, and generate mapping of `feature_id` to
+/// `EnrolledFeatureConfig` structs.
+pub fn map_features(
+    enrollments: &[ExperimentEnrollment],
+    experiments: &HashMap<String, &Experiment>,
+) -> HashMap<String, EnrolledFeatureConfig> {
+    let mut map = HashMap::with_capacity(enrollments.len());
+    for e in enrollments {
+        if let Some(feature) = get_feature_config(e, experiments) {
+            map.insert(feature.feature_id.clone(), feature);
+        }
+    }
+    map
```

Comment:
I wonder if there is a more Rusty way of doing this..  something like:
```rust
let m = enrollments.iter().filter_map(|e| get_feature_config(e, experiments).is_some()).collect();
```

Reply:
TIL!

Reply:
@travis79 Is something like this what you had in mind?
```rust
    for enrolled_feature_config in enrollments
        .iter()
        .filter_map(|e| get_feature_config(e, experiments))
    {
        map.insert(
            enrolled_feature_config.feature_id.clone(),
            enrolled_feature_config,
        );
    }
```

Reply:
The enrollments.iter() get's the iterator which allows you to both filter on `is_some()` and map to the result of the call to `get_feature_config`, and then `collect()` them back into a collection again.

Reply:
I don't have your branch checked out, but here is the `map_enrollments` funtion in `enrollment.rs` with the same simplification:

This:
```rust
fn map_enrollments(enrollments: &[ExperimentEnrollment]) -> HashMap<String, &ExperimentEnrollment> {
    let mut map_enrollments = HashMap::with_capacity(enrollments.len());
    for e in enrollments {
        map_enrollments.insert(e.slug.clone(), e);
    }
    map_enrollments
}
```

Becomes this:
```rust
fn map_enrollments(enrollments: &[ExperimentEnrollment]) -> HashMap<String, &ExperimentEnrollment> {
    enrollments.iter().map(|e| (e.slug.clone(), e)).collect()
}
```

I've suggested you use `filter_map`, but the concept is similar.

Reply:
```rust
pub fn map_features(
    enrollments: &[ExperimentEnrollment],
    experiments: &HashMap<String, &Experiment>,
) -> HashMap<String, EnrolledFeatureConfig> {
    enrollments.iter().filter_map(|e| get_feature_config(e, experiments).is_some()).collect()
}
```

This is more what I was suggesting ^^

Reply:
(checking out your branch now to try a for-real example of what I am thinking...)

Reply:
OK, that's super helpful -- the trick was using the tuple for the key/value; thanks!

Reply:
Nevermind, this is turning out to be more of a rabbit hole than I thought.  Digging the feature_id out complicates things a bit.

Reply:
Yeah, I was just starting to conclude the same thing.   Are you OK with me leaving it as the last thing I proposed in this thread?

Reply:
Here it would need to be something like this (which I'm not sure is more readable or not):
```rust
/// Take a list of enrollments and a map of experiments, and generate mapping of `feature_id` to
/// `EnrolledFeatureConfig` structs.
pub fn map_features(
    enrollments: &[ExperimentEnrollment],
    experiments: &HashMap<String, &Experiment>,
) -> HashMap<String, EnrolledFeatureConfig> {
    enrollments.iter().filter_map(|e|
        get_feature_config(e, experiments)
    ).map(|config|
        (config.feature_id.clone(), config)
    ).collect()
}
```

Reply:
I'll leave it up to you whether you want to use this or not.